From 60b5027f1b7106af05b407389b612c275e64960d Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Thu, 23 Aug 2018 18:18:05 +0900
Subject: [PATCH 293/909] drm: rcar-du: Add HDSR and VDSR register value check

commit 8cfac793337cb17c0b21e46cfdac8c34811e7553 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

It is prohibited to set a value less than 1 to VDS bit of VDSR
register and HDS bit of HDSR by the H/W specification,
so this patch adds judgement to prohibit setting a value less than 1.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/rcar-du/rcar_du_crtc.c | 31 ++++++++++++++++++++++++++
 1 file changed, 31 insertions(+)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
index a8eb15fd13ac..309c9b34e65c 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
@@ -686,11 +686,42 @@ static void rcar_du_crtc_atomic_flush(struct drm_crtc *crtc,
 		rcar_du_vsp_atomic_flush(rcrtc);
 }
 
+static bool rcar_du_crtc_mode_fixup(struct drm_crtc *crtc,
+				    const struct drm_display_mode *mode,
+				    struct drm_display_mode *adjusted_mode)
+{
+	struct rcar_du_crtc *rcrtc = to_rcar_crtc(crtc);
+	struct rcar_du_device *rcdu = rcrtc->group->dev;
+	int vdsr_reg = mode->crtc_vtotal - mode->crtc_vsync_end - 2;
+	int hdsr_reg = mode->htotal - mode->hsync_start - 19;
+
+	/*
+	 * It is prohibited to set a value less than 1 to VDSR and HDSR
+	 * register by the H/W specification.
+	 */
+	if (vdsr_reg < 1) {
+		dev_err(rcdu->dev,
+			"setting value (%d) to VDSR register is invalid.\n",
+			vdsr_reg);
+		return false;
+	}
+
+	if (hdsr_reg < 1) {
+		dev_err(rcdu->dev,
+			"setting value (%d) to HDSR register is invalid.\n",
+			hdsr_reg);
+		return false;
+	}
+
+	return true;
+}
+
 static const struct drm_crtc_helper_funcs crtc_helper_funcs = {
 	.atomic_begin = rcar_du_crtc_atomic_begin,
 	.atomic_flush = rcar_du_crtc_atomic_flush,
 	.atomic_enable = rcar_du_crtc_atomic_enable,
 	.atomic_disable = rcar_du_crtc_atomic_disable,
+	.mode_fixup = rcar_du_crtc_mode_fixup,
 };
 
 static struct drm_crtc_state *
-- 
2.17.1

